File: python-service/scripts/ingest_data.py
import os
import time
from pathlib import Path
from pymongo import MongoClient
from voyageai import Client as VoyageClient
from dotenv import load_dotenv

load_dotenv()

DATA_DIR = Path(__file__).parent.parent / "data"
voyage_client = VoyageClient(api_key=os.getenv("VOYAGE_API_KEY"))


def chunk_text(text: str, chunk_size: int = 2400, overlap: int = 200) -> list[str]:
    """Split text into overlapping chunks."""
    chunks = []
    start = 0
    while start < len(text):
        end = min(start + chunk_size, len(text))
        chunk = text[start:end].strip()
        if len(chunk) > 100:
            chunks.append(chunk)
        start += chunk_size - overlap
    return chunks


def topic_from_filename(filename: str) -> str:
    return filename.replace(".md", "").replace("-", " ").replace("_", " ").title()


def ingest_data():
    print("🚀 Starting data ingestion...")

    client = MongoClient(os.getenv("MONGODB_URI"))
    db = client["ai_tutor"]
    collection = db["placement_docs"]

    # Clear existing data
    collection.delete_many({})
    print("🗑️  Cleared existing placement_docs collection")

    md_files = list(DATA_DIR.glob("*.md"))
    print(f"📂 Found {len(md_files)} markdown files")

    total = 0
    for md_file in md_files:
        topic = topic_from_filename(md_file.name)
        text = md_file.read_text(encoding="utf-8")
        chunks = chunk_text(text)
        print(f"\n📄 Processing '{topic}' → {len(chunks)} chunks")

        for i, chunk in enumerate(chunks):
            try:
                embed_response = voyage_client.embed(
                    texts=[chunk],
                    model="voyage-3-lite",
                )
                embedding = embed_response.embeddings[0]

                collection.insert_one({
                    "topic": topic,
                    "subtopic": f"{topic} - Part {i + 1}",
                    "content": chunk,
                    "embedding": embedding,
                })
                total += 1
                print(f"  ✅ Chunk {i + 1}/{len(chunks)} stored")
                
                # Voyage AI free tier without CC has a 3 RPM (Requests Per Minute) limit.
                # We must sleep for at least 20 seconds between requests.
                if total < sum(len(chunk_text(Path(f).read_text(encoding="utf-8"))) for f in md_files):
                    print("  ⏳ Sleeping 21s to respect Voyage AI free tier 3 RPM limit...")
                    time.sleep(21)

            except Exception as e:
                print(f"  ❌ Error on chunk {i + 1}: {e}")

    print(f"\n✅ Done! {total} chunks stored in MongoDB Atlas")
    print("\n📌 Now create a Vector Search Index in MongoDB Atlas UI:")
    print("  Collection: ai_tutor.placement_docs")
    print("  Index name: vector_index")
    print("""  Config:
{
  "fields": [{
    "type": "vector",
    "path": "embedding",
    "numDimensions": 512,
    "similarity": "cosine"
  }]
}""")
    client.close()


if __name__ == "__main__":
    ingest_data()

File: backend/scripts/ingestData.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { MongoClient } from "mongodb";
import VoyageAI from "voyageai";
import dotenv from "dotenv";
dotenv.config();

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DATA_DIR = path.join(__dirname, "../data");

const voyage = new VoyageAI({ apiKey: process.env.VOYAGE_API_KEY });

// Chunk text into segments of ~600 tokens (~2400 chars)
function chunkText(text, chunkSize = 2400, overlap = 200) {
    const chunks = [];
    let start = 0;
    while (start < text.length) {
        const end = Math.min(start + chunkSize, text.length);
        chunks.push(text.slice(start, end).trim());
        start += chunkSize - overlap;
    }
    return chunks.filter((c) => c.length > 100);
}

// Extract topic from filename
function topicFromFilename(filename) {
    return filename
        .replace(/\.md$/, "")
        .replace(/[-_]/g, " ")
        .replace(/\b\w/g, (c) => c.toUpperCase());
}

// Sleep helper to avoid rate limits
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function ingestData() {
    console.log("🚀 Starting data ingestion...");

    const client = new MongoClient(process.env.MONGODB_URI);
    await client.connect();
    const db = client.db("ai_tutor");
    const collection = db.collection("placement_docs");

    // Clear existing data
    await collection.deleteMany({});
    console.log("🗑️  Cleared existing placement_docs collection");

    const files = fs.readdirSync(DATA_DIR).filter((f) => f.endsWith(".md"));
    console.log(`📂 Found ${files.length} markdown files`);

    let totalChunks = 0;

    for (const file of files) {
        const topic = topicFromFilename(file);
        const text = fs.readFileSync(path.join(DATA_DIR, file), "utf-8");
        const chunks = chunkText(text);

        console.log(`\n📄 Processing "${topic}" → ${chunks.length} chunks`);

        for (let i = 0; i < chunks.length; i++) {
            try {
                // Generate embedding for this chunk
                const embeddingResponse = await voyage.embed({
                    input: chunks[i],
                    model: "voyage-3-lite",
                });
                const embedding = embeddingResponse.data[0].embedding;

                // Store in MongoDB
                await collection.insertOne({
                    topic,
                    subtopic: `${topic} - Part ${i + 1}`,
                    content: chunks[i],
                    embedding,
                    createdAt: new Date(),
                });

                totalChunks++;
                console.log(`  ✅ Chunk ${i + 1}/${chunks.length} embedded and stored`);

                // Rate limit: 100ms between requests
                await sleep(100);
            } catch (error) {
                console.error(`  ❌ Error on chunk ${i + 1}:`, error.message);
            }
        }
    }

    console.log(`\n✅ Ingestion complete! Stored ${totalChunks} chunks in MongoDB Atlas`);
    console.log("\n📌 Next step: Create a Vector Search Index in MongoDB Atlas:");
    console.log("  1. Go to your Atlas cluster → Search → Create Search Index");
    console.log('  2. Choose "Atlas Vector Search" → JSON editor');
    console.log("  3. Collection: ai_tutor.placement_docs");
    console.log(`  4. Paste this config:
{
  "fields": [
    {
      "type": "vector",
      "path": "embedding",
      "numDimensions": 512,
      "similarity": "cosine"
    }
  ]
}
  5. Name it: vector_index`);

    await client.close();
}

ingestData().catch(console.error);
